File: Luna/JSLoader/Services/StremioModels.swift
//
//  StremioModels.swift
//  Luna
//
//  Created by Soupy on 2026.
//

import Foundation
import CoreData

// MARK: - Stremio Manifest

struct StremioManifest: Codable {
    let id: String
    let name: String
    let description: String?
    let version: String?
    let logo: String?
    let types: [String]?
    let resources: [StremioResource]?
    let idPrefixes: [String]?
    let catalogs: [StremioCatalog]?
    let behaviorHints: StremioManifestBehaviorHints?

    enum CodingKeys: String, CodingKey {
        case id, name, description, version, logo, types, resources, idPrefixes, catalogs, behaviorHints
    }

    /// Whether this addon supports a given ID prefix (e.g. "tt", "tmdb:", "kitsu:")
    func supportsPrefix(_ prefix: String) -> Bool {
        guard let prefixes = idPrefixes, !prefixes.isEmpty else { return true }
        return prefixes.contains(where: { prefix.hasPrefix($0) })
    }

    /// Whether this addon supports the "stream" resource
    var supportsStreams: Bool {
        guard let resources = resources else { return false }
        return resources.contains { $0.isStream }
    }

    /// Whether this addon supports the "subtitles" resource
    var supportsSubtitles: Bool {
        guard let resources = resources else { return false }
        return resources.contains { $0.isSubtitles }
    }

    var supportsMeta: Bool {
        guard let resources = resources else { return false }
        return resources.contains { $0.isMeta }
    }

    var searchableCatalogs: [StremioCatalog] {
        catalogs?.filter { $0.canSearchWithQueryOnly } ?? []
    }

    var streamIdPrefixes: [String]? {
        let resourcePrefixes = resources?
            .flatMap { $0.idPrefixes(for: "stream") }
            .filter { !$0.isEmpty } ?? []
        return resourcePrefixes.isEmpty ? idPrefixes : resourcePrefixes
    }
}

struct StremioManifestBehaviorHints: Codable {
    let configurable: Bool?
    let configurationRequired: Bool?
}

// MARK: - Resource (can be a string or an object)

enum StremioResource: Codable {
    case simple(String)
    case detailed(StremioResourceDetail)

    var isStream: Bool {
        switch self {
        case .simple(let name): return name == "stream"
        case .detailed(let detail): return detail.name == "stream"
        }
    }

    var isSubtitles: Bool {
        switch self {
        case .simple(let name): return name == "subtitles"
        case .detailed(let detail): return detail.name == "subtitles"
        }
    }

    var isMeta: Bool {
        switch self {
        case .simple(let name): return name == "meta"
        case .detailed(let detail): return detail.name == "meta"
        }
    }

    func idPrefixes(for resourceName: String) -> [String] {
        switch self {
        case .simple:
            return []
        case .detailed(let detail):
            return detail.name == resourceName ? (detail.idPrefixes ?? []) : []
        }
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let string = try? container.decode(String.self) {
            self = .simple(string)
        } else {
            let detail = try container.decode(StremioResourceDetail.self)
            self = .detailed(detail)
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .simple(let name):
            try container.encode(name)
        case .detailed(let detail):
            try container.encode(detail)
        }
    }
}

struct StremioResourceDetail: Codable {
    let name: String
    let types: [String]?
    let idPrefixes: [String]?
}

// MARK: - Catalog and Meta Responses

struct StremioCatalog: Codable, Hashable {
    let type: String
    let id: String
    let name: String?
    let extra: [StremioCatalogExtra]?

    var supportsSearch: Bool {
        extra?.contains { $0.name == "search" } ?? false
    }

    var canSearchWithQueryOnly: Bool {
        guard supportsSearch else { return false }
        return extra?.allSatisfy { extra in
            extra.isRequired != true || extra.name == "search"
        } ?? true
    }

    func supportsType(_ requestedType: String) -> Bool {
        type == requestedType || (requestedType == "series" && type == "tv")
    }
}

struct StremioCatalogExtra: Codable, Hashable {
    let name: String
    let isRequired: Bool?
    let options: [String]?
    let optionsLimit: Int?

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let name = try? container.decode(String.self) {
            self.name = name
            self.isRequired = nil
            self.options = nil
            self.optionsLimit = nil
            return
        }

        let keyed = try decoder.container(keyedBy: CodingKeys.self)
        name = (try? keyed.decode(String.self, forKey: .name)) ?? ""
        isRequired = try? keyed.decodeIfPresent(Bool.self, forKey: .isRequired)
        options = try? keyed.decodeIfPresent([String].self, forKey: .options)
        optionsLimit = try? keyed.decodeIfPresent(Int.self, forKey: .optionsLimit)
    }
}

struct StremioCatalogResponse: Codable {
    let metas: [StremioMetaPreview]

    enum CodingKeys: String, CodingKey {
        case metas
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        metas = Self.decodeLossyArray(from: container, forKey: .metas)
    }

    private static func decodeLossyArray(
        from container: KeyedDecodingContainer<CodingKeys>,
        forKey key: CodingKeys
    ) -> [StremioMetaPreview] {
        guard var unkeyedContainer = try? container.nestedUnkeyedContainer(forKey: key) else {
            return []
        }

        var decoded = [StremioMetaPreview]()
        while !unkeyedContainer.isAtEnd {
            if let meta = try? unkeyedContainer.decode(StremioMetaPreview.self) {
                decoded.append(meta)
            } else {
                _ = try? unkeyedContainer.decode(AnyCodable.self)
            }
        }
        return decoded
    }
}

struct StremioMetaResponse: Codable {
    let meta: StremioMetaPreview?

    enum CodingKeys: String, CodingKey {
        case meta
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let single = try? container.decodeIfPresent(StremioMetaPreview.self, forKey: .meta) {
            meta = single
        } else if let array = try? container.decodeIfPresent([StremioMetaPreview].self, forKey: .meta) {
            meta = array.first
        } else {
            meta = nil
        }
    }
}

struct StremioMetaPreview: Codable, Identifiable, Hashable {
    let id: String
    let type: String?
    let name: String
    let poster: String?
    let description: String?
    let releaseInfo: String?
    let released: String?
    let videos: [StremioVideo]?
    let behaviorHints: StremioMetaBehaviorHints?

    enum CodingKeys: String, CodingKey {
        case id, type, name, poster, description, releaseInfo, released, videos, behaviorHints
    }
}

struct StremioMetaBehaviorHints: Codable, Hashable {
    let defaultVideoId: String?
}

struct StremioVideo: Codable, Identifiable, Hashable {
    let id: String
    let title: String?
    let released: String?
    let season: Int?
    let episode: Int?
    let streams: [StremioStream]?

    enum CodingKeys: String, CodingKey {
        case id, title, released, season, episode, streams
    }
}

// MARK: - Stream Response

struct StremioStreamResponse: Codable {
    let streams: [StremioStream]?

    enum CodingKeys: String, CodingKey {
        case streams
    }

    /// Lossy decoding: skips individual streams that fail to decode instead of
    /// dropping the entire array (the default Codable behaviour).
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // Try decoding each element individually; skip failures
        if var unkeyedContainer = try? container.nestedUnkeyedContainer(forKey: .streams) {
            var decoded = [StremioStream]()
            while !unkeyedContainer.isAtEnd {
                if let stream = try? unkeyedContainer.decode(StremioStream.self) {
                    decoded.append(stream)
                } else {
                    // Skip the bad element so the container advances
                    _ = try? unkeyedContainer.decode(AnyCodable.self)
                }
            }
            streams = decoded.isEmpty ? nil : decoded
        } else {
            streams = nil
        }
    }
}

struct StremioSubtitleResponse: Codable, Sendable {
    let subtitles: [StremioSubtitle]?

    enum CodingKeys: String, CodingKey {
        case subtitles
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if var unkeyedContainer = try? container.nestedUnkeyedContainer(forKey: .subtitles) {
            var decoded = [StremioSubtitle]()
            while !unkeyedContainer.isAtEnd {
                if let subtitle = try? unkeyedContainer.decode(StremioSubtitle.self) {
                    decoded.append(subtitle)
                } else {
                    _ = try? unkeyedContainer.decode(AnyCodable.self)
                }
            }
            subtitles = decoded.isEmpty ? nil : decoded
        } else {
            subtitles = nil
        }
    }
}

/// Throwaway type used only to advance the unkeyed container past a bad element.
private struct AnyCodable: Codable {
    init(from decoder: Decoder) throws {
        _ = try? decoder.singleValueContainer().decode(String.self)
    }
    func encode(to encoder: Encoder) throws {}
}

struct StremioStream: Codable, Identifiable, Hashable {
    let id: String

    let url: String?
    let infoHash: String?
    let title: String?
    let name: String?
    let description: String?
    let behaviorHints: StremioStreamBehaviorHints?
    let subtitles: [StremioSubtitle]?

    enum CodingKeys: String, CodingKey {
        case url, infoHash, title, name, description, behaviorHints, subtitles
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        url = try container.decodeIfPresent(String.self, forKey: .url)
        infoHash = try container.decodeIfPresent(String.self, forKey: .infoHash)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        name = try container.decodeIfPresent(String.self, forKey: .name)
        description = try container.decodeIfPresent(String.self, forKey: .description)
        // Use try? so unexpected shapes don't kill the whole stream
        behaviorHints = try? container.decodeIfPresent(StremioStreamBehaviorHints.self, forKey: .behaviorHints)
        subtitles = try? container.decodeIfPresent([StremioSubtitle].self, forKey: .subtitles)
        id = url ?? infoHash ?? UUID().uuidString
    }

    /// Whether this stream is a direct HTTP(S) link (safe, no torrent)
    var isDirectHTTP: Bool {
        guard let url = url, !url.isEmpty else { return false }
        let lower = url.lowercased()
        return lower.hasPrefix("http://") || lower.hasPrefix("https://")
    }

    /// Display name for the stream (prefers name, falls back to title)
    var displayName: String {
        if let name = name, !name.isEmpty { return name }
        if let title = title, !title.isEmpty { return title }
        return "Stream"
    }

    /// Extracts proxy headers from behaviorHints if available
    var proxyHeaders: [String: String]? {
        behaviorHints?.proxyHeaders?.request
    }
}

struct StremioStreamBehaviorHints: Codable, Hashable {
    let notWebReady: Bool?
    let bingeGroup: String?
    let proxyHeaders: StremioProxyHeaders?
    let filename: String?

    enum CodingKeys: String, CodingKey {
        case notWebReady, bingeGroup, proxyHeaders, filename
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // notWebReady can arrive as Bool, Int, or String from various addons
        if let b = try? container.decodeIfPresent(Bool.self, forKey: .notWebReady) {
            notWebReady = b
        } else if let i = try? container.decodeIfPresent(Int.self, forKey: .notWebReady) {
            notWebReady = i != 0
        } else {
            notWebReady = nil
        }
        bingeGroup = try? container.decodeIfPresent(String.self, forKey: .bingeGroup)
        proxyHeaders = try? container.decodeIfPresent(StremioProxyHeaders.self, forKey: .proxyHeaders)
        filename = try? container.decodeIfPresent(String.self, forKey: .filename)
    }
}

struct StremioProxyHeaders: Codable, Hashable {
    let request: [String: String]?
}

struct StremioSubtitle: Codable, Sendable, Hashable {
    let id: String?
    let url: String?
    let lang: String?
    let name: String?
    let title: String?

    enum CodingKeys: String, CodingKey {
        case id, url, lang, name, title
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // Some addons return subtitle id as an integer
        if let s = try? container.decodeIfPresent(String.self, forKey: .id) {
            id = s
        } else if let i = try? container.decodeIfPresent(Int.self, forKey: .id) {
            id = String(i)
        } else {
            id = nil
        }
        url = try? container.decodeIfPresent(String.self, forKey: .url)
        lang = try? container.decodeIfPresent(String.self, forKey: .lang)
        name = try? container.decodeIfPresent(String.self, forKey: .name)
        title = try? container.decodeIfPresent(String.self, forKey: .title)
    }

    var displayName: String {
        if let name, !name.isEmpty { return name }
        if let title, !title.isEmpty { return title }
        if let lang, !lang.isEmpty { return lang.uppercased() }
        return id ?? "OpenSubtitles"
    }
}

// MARK: - Stremio Addon Model (persisted)

struct StremioAddon: Identifiable, Hashable {
    let id: UUID
    let configuredURL: String
    let manifest: StremioManifest
    let isActive: Bool
    let sortIndex: Int64

    static func == (lhs: StremioAddon, rhs: StremioAddon) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - StremioAddonEntity (CoreData)

@objc(StremioAddonEntity)
public class StremioAddonEntity: NSManagedObject { }

extension StremioAddonEntity {
    @nonobjc public class func fetchRequest() -> NSFetchRequest<StremioAddonEntity> {
        return NSFetchRequest<StremioAddonEntity>(entityName: "StremioAddonEntity")
    }

    @NSManaged public var id: UUID?
    @NSManaged public var configuredURL: String?
    @NSManaged public var manifestJSON: String?
    @NSManaged public var isActive: Bool
    @NSManaged public var sortIndex: Int64

    override public func awakeFromInsert() {
        super.awakeFromInsert()
        if id == nil {
            let temp = UUID()
            id = temp
            Logger.shared.log("Added empty StremioAddonEntity: \(temp)", type: "Stremio")
        }
    }
}

extension StremioAddonEntity: Identifiable { }

extension StremioAddonEntity {
    var asModel: StremioAddon? {
        guard
            let id = self.id,
            let configuredURL = self.configuredURL,
            let manifestJSON = self.manifestJSON,
            let data = manifestJSON.data(using: .utf8)
        else {
            return nil
        }

        do {
            let manifest = try JSONDecoder().decode(StremioManifest.self, from: data)
            return StremioAddon(
                id: id,
                configuredURL: configuredURL,
                manifest: manifest,
                isActive: isActive,
                sortIndex: sortIndex
            )
        } catch {
            Logger.shared.log("Failed to decode StremioManifest for \(id.uuidString): \(error.localizedDescription)", type: "Stremio")
            return nil
        }
    }
}

File: android/core/model/src/main/kotlin/dev/soupy/eclipse/android/core/model/StremioModels.kt
package dev.soupy.eclipse.android.core.model
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder

private val qualityPatterns = listOf(
    "2160p" to 1.00,
    "4k" to 1.00,
    "1080p" to 0.90,
    "720p" to 0.72,
    "480p" to 0.48,
    "cam" to -0.35,
    "hdcam" to -0.35,
    "telesync" to -0.30,
    " ts " to -0.30,
)

@Serializable
data class StremioManifestBehaviorHints(
    val configurable: Boolean = false,
    @SerialName("configurationRequired") val configurationRequired: Boolean = false,
)

@Serializable(with = StremioResourceDescriptorSerializer::class)
data class StremioResourceDescriptor(
    val name: String = "",
    val types: List<String> = emptyList(),
    @SerialName("idPrefixes") val idPrefixes: List<String> = emptyList(),
)

@Serializable
data class StremioManifest(
    val id: String = "",
    val version: String = "",
    val name: String = "",
    val description: String? = null,
    @SerialName("logo") val logoUrl: String? = null,
    val background: String? = null,
    val resources: List<StremioResourceDescriptor> = emptyList(),
    @SerialName("idPrefixes") val idPrefixes: List<String> = emptyList(),
    val types: List<String> = emptyList(),
    val catalogs: List<StremioCatalog> = emptyList(),
    @SerialName("behaviorHints") val behaviorHints: StremioManifestBehaviorHints = StremioManifestBehaviorHints(),
)

@Serializable
data class StremioCatalog(
    val type: String = "",
    val id: String = "",
    val name: String? = null,
    val extra: List<StremioCatalogExtra> = emptyList(),
) {
    val supportsSearch: Boolean
        get() = extra.any { catalogExtra -> catalogExtra.name == "search" }

    val canSearchWithQueryOnly: Boolean
        get() = supportsSearch && extra.all { catalogExtra ->
            !catalogExtra.isRequired || catalogExtra.name == "search"
        }

    fun supportsType(requestedType: String): Boolean =
        type == requestedType || (requestedType == "series" && type == "tv")
}

@Serializable(with = StremioCatalogExtraSerializer::class)
data class StremioCatalogExtra(
    val name: String = "",
    @SerialName("isRequired") val isRequired: Boolean = false,
    val options: List<String> = emptyList(),
    @SerialName("optionsLimit") val optionsLimit: Int? = null,
)

@Serializable
data class StremioProxyHeaders(
    val request: Map<String, String> = emptyMap(),
    val response: Map<String, String> = emptyMap(),
)

@Serializable(with = StremioSubtitleSerializer::class)
data class StremioSubtitle(
    val id: String? = null,
    val lang: String? = null,
    val label: String? = null,
    val url: String? = null,
    val name: String? = null,
    val title: String? = null,
)

@Serializable
data class StremioStreamBehaviorHints(
    @SerialName("bingeGroup") val bingeGroup: String? = null,
    @SerialName("filename") val filename: String? = null,
    @SerialName("notWebReady") val notWebReady: Boolean = false,
    @SerialName("proxyHeaders") val proxyHeaders: StremioProxyHeaders? = null,
)

@Serializable
data class StremioStream(
    val name: String? = null,
    val title: String? = null,
    val description: String? = null,
    val url: String? = null,
    @SerialName("ytId") val ytId: String? = null,
    val infoHash: String? = null,
    val fileIdx: Int? = null,
    val subtitles: List<StremioSubtitle> = emptyList(),
    @SerialName("behaviorHints") val behaviorHints: StremioStreamBehaviorHints? = null,
)

@Serializable
data class StremioStreamResponse(
    val streams: List<StremioStream> = emptyList(),
)

@Serializable
data class StremioSubtitleResponse(
    val subtitles: List<StremioSubtitle> = emptyList(),
)

@Serializable
data class StremioCatalogResponse(
    val metas: List<StremioMetaPreview> = emptyList(),
)

@Serializable(with = StremioMetaResponseSerializer::class)
data class StremioMetaResponse(
    val meta: StremioMetaPreview? = null,
)

@Serializable
data class StremioMetaPreview(
    val id: String = "",
    val type: String? = null,
    val name: String = "",
    val poster: String? = null,
    val description: String? = null,
    @SerialName("releaseInfo") val releaseInfo: String? = null,
    val released: String? = null,
    val videos: List<StremioVideo> = emptyList(),
    @SerialName("behaviorHints") val behaviorHints: StremioMetaBehaviorHints? = null,
)

@Serializable
data class StremioMetaBehaviorHints(
    @SerialName("defaultVideoId") val defaultVideoId: String? = null,
)

@Serializable
data class StremioVideo(
    val id: String = "",
    val title: String? = null,
    val released: String? = null,
    val season: Int? = null,
    val episode: Int? = null,
    val streams: List<StremioStream> = emptyList(),
)

@Serializable
data class StremioAddon(
    val transportUrl: String,
    val manifest: StremioManifest,
    val enabled: Boolean = true,
    val sortIndex: Int = 0,
)

@Serializable
data class StremioContentIdRequest(
    val tmdbId: Int,
    val imdbId: String? = null,
    val type: String,
    val season: Int? = null,
    val episode: Int? = null,
    val anilistId: Int? = null,
)

val StremioStream.isDirectHttp: Boolean
    get() = url?.startsWith("http://") == true || url?.startsWith("https://") == true

val StremioStream.isTorrentLike: Boolean
    get() = !infoHash.isNullOrBlank() ||
        url?.startsWith("magnet:", ignoreCase = true) == true ||
        url?.contains("btih:", ignoreCase = true) == true ||
        url
            ?.substringBefore('?')
            ?.substringBefore('#')
            ?.endsWith(".torrent", ignoreCase = true) == true

fun StremioManifest.supportsResource(name: String): Boolean =
    resources.any { resource -> resource.name.equals(name, ignoreCase = true) }

val StremioManifest.searchableCatalogs: List<StremioCatalog>
    get() = catalogs.filter(StremioCatalog::canSearchWithQueryOnly)

fun StremioManifest.buildContentId(
    request: StremioContentIdRequest,
    resourceName: String = "stream",
): String? = buildContentIds(request = request, resourceName = resourceName).firstOrNull()

fun StremioManifest.buildContentIds(
    request: StremioContentIdRequest,
    resourceName: String = "stream",
): List<String> {
    val resourcePrefixes = resources
        .filter { resource -> resource.name.equals(resourceName, ignoreCase = true) }
        .flatMap(StremioResourceDescriptor::idPrefixes)
        .filter(String::isNotBlank)
    val prefixes = resourcePrefixes.ifEmpty { idPrefixes }.map { prefix -> prefix.lowercase() }
    val supportsAny = prefixes.isEmpty()
    val supportsImdb = supportsAny || prefixes.any { prefix ->
        prefix == "tt" || prefix.startsWith("tt") || prefix == "imdb" || prefix == "imdb:"
    }
    val supportsImdbNamespace = prefixes.any { prefix -> prefix == "imdb:" }
    val supportsTmdb = supportsAny || prefixes.any { prefix ->
        prefix == "tmdb" || prefix.startsWith("tmdb:")
    }
    val supportsAniList = supportsAny || prefixes.any { prefix ->
        prefix == "anilist" || prefix == "anilist:"
    }
    val candidates = mutableListOf<String>()

    if (supportsImdb) {
        val imdb = request.imdbId?.takeIf { it.isNotBlank() }?.let { value ->
            if (value.startsWith("tt")) value else "tt$value"
        }
        if (imdb != null) {
            val imdbId = if (request.type == "series" && request.season != null && request.episode != null) {
                "$imdb:${request.season}:${request.episode}"
            } else {
                imdb
            }
            candidates += imdbId

            if (supportsImdbNamespace) {
                candidates += if (request.type == "series" && request.season != null && request.episode != null) {
                    "imdb:$imdb:${request.season}:${request.episode}"
                } else {
                    "imdb:$imdb"
                }
            }
        }
    }

    if (supportsTmdb) {
        candidates += if (request.type == "series" && request.season != null && request.episode != null) {
            "tmdb:${request.tmdbId}:${request.season}:${request.episode}"
        } else {
            "tmdb:${request.tmdbId}"
        }
    }

    if (supportsAniList && request.anilistId != null) {
        candidates += if (request.type == "series" && request.season != null && request.episode != null) {
            "anilist:${request.anilistId}:${request.season}:${request.episode}"
        } else {
            "anilist:${request.anilistId}"
        }
    }

    return candidates.distinct()
}

val StremioSubtitle.displayLabel: String
    get() = listOfNotNull(label, name, title, lang?.uppercase(), id)
        .firstOrNull { it.isNotBlank() }
        ?: "OpenSubtitles"

object StremioResourceDescriptorSerializer : KSerializer<StremioResourceDescriptor> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("StremioResourceDescriptor") {
        element<String>("name")
        element<List<String>>("types", isOptional = true)
        element<List<String>>("idPrefixes", isOptional = true)
    }

    override fun deserialize(decoder: Decoder): StremioResourceDescriptor {
        val input = decoder as? JsonDecoder ?: error("StremioResourceDescriptor requires JSON decoding")
        return when (val element = input.decodeJsonElement()) {
            is JsonPrimitive -> StremioResourceDescriptor(name = element.contentOrNull.orEmpty())
            else -> input.json.decodeFromJsonElement<StremioResourceDescriptorSurrogate>(element).toResourceDescriptor()
        }
    }

    override fun serialize(encoder: Encoder, value: StremioResourceDescriptor) {
        val output = encoder as? JsonEncoder ?: error("StremioResourceDescriptor requires JSON encoding")
        output.encodeJsonElement(
            output.json.encodeToJsonElement(
                StremioResourceDescriptorSurrogate(
                    name = value.name,
                    types = value.types,
                    idPrefixes = value.idPrefixes,
                ),
            ),
        )
    }
}

object StremioSubtitleSerializer : KSerializer<StremioSubtitle> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("StremioSubtitle") {
        element<String>("id", isOptional = true)
        element<String>("lang", isOptional = true)
        element<String>("label", isOptional = true)
        element<String>("url", isOptional = true)
        element<String>("name", isOptional = true)
        element<String>("title", isOptional = true)
    }

    override fun deserialize(decoder: Decoder): StremioSubtitle {
        val input = decoder as? JsonDecoder ?: error("StremioSubtitle requires JSON decoding")
        val surrogate = input.json.decodeFromJsonElement<StremioSubtitleSurrogate>(input.decodeJsonElement())
        return StremioSubtitle(
            id = surrogate.id?.jsonPrimitive?.contentOrNull,
            lang = surrogate.lang,
            label = surrogate.label,
            url = surrogate.url,
            name = surrogate.name,
            title = surrogate.title,
        )
    }

    override fun serialize(encoder: Encoder, value: StremioSubtitle) {
        val output = encoder as? JsonEncoder ?: error("StremioSubtitle requires JSON encoding")
        output.encodeJsonElement(
            output.json.encodeToJsonElement(
                StremioSubtitleSurrogate(
                    id = value.id?.let(::JsonPrimitive),
                    lang = value.lang,
                    label = value.label,
                    url = value.url,
                    name = value.name,
                    title = value.title,
                ),
            ),
        )
    }
}

object StremioCatalogExtraSerializer : KSerializer<StremioCatalogExtra> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("StremioCatalogExtra") {
        element<String>("name")
        element<Boolean>("isRequired", isOptional = true)
        element<List<String>>("options", isOptional = true)
        element<Int>("optionsLimit", isOptional = true)
    }

    override fun deserialize(decoder: Decoder): StremioCatalogExtra {
        val input = decoder as? JsonDecoder ?: error("StremioCatalogExtra requires JSON decoding")
        return when (val element = input.decodeJsonElement()) {
            is JsonPrimitive -> StremioCatalogExtra(name = element.contentOrNull.orEmpty())
            else -> input.json.decodeFromJsonElement<StremioCatalogExtraSurrogate>(element).toCatalogExtra()
        }
    }

    override fun serialize(encoder: Encoder, value: StremioCatalogExtra) {
        val output = encoder as? JsonEncoder ?: error("StremioCatalogExtra requires JSON encoding")
        output.encodeJsonElement(
            output.json.encodeToJsonElement(
                StremioCatalogExtraSurrogate(
                    name = value.name,
                    isRequired = value.isRequired,
                    options = value.options,
                    optionsLimit = value.optionsLimit,
                ),
            ),
        )
    }
}

object StremioMetaResponseSerializer : KSerializer<StremioMetaResponse> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("StremioMetaResponse") {
        element<StremioMetaPreview>("meta", isOptional = true)
    }

    override fun deserialize(decoder: Decoder): StremioMetaResponse {
        val input = decoder as? JsonDecoder ?: error("StremioMetaResponse requires JSON decoding")
        val root = input.decodeJsonElement() as? JsonObject ?: return StremioMetaResponse()
        val metaElement = root["meta"] ?: return StremioMetaResponse()
        val meta = when (metaElement) {
            is JsonArray -> metaElement.firstOrNull()?.let { element ->
                runCatching { input.json.decodeFromJsonElement<StremioMetaPreview>(element) }.getOrNull()
            }
            is JsonObject -> runCatching {
                input.json.decodeFromJsonElement<StremioMetaPreview>(metaElement)
            }.getOrNull()
            else -> null
        }
        return StremioMetaResponse(meta = meta)
    }

    override fun serialize(encoder: Encoder, value: StremioMetaResponse) {
        val output = encoder as? JsonEncoder ?: error("StremioMetaResponse requires JSON encoding")
        output.encodeJsonElement(output.json.encodeToJsonElement(value.meta?.let { mapOf("meta" to it) } ?: emptyMap()))
    }
}

@Serializable
private data class StremioResourceDescriptorSurrogate(
    val name: String = "",
    val types: List<String> = emptyList(),
    @SerialName("idPrefixes") val idPrefixes: List<String> = emptyList(),
) {
    fun toResourceDescriptor(): StremioResourceDescriptor = StremioResourceDescriptor(
        name = name,
        types = types,
        idPrefixes = idPrefixes,
    )
}

@Serializable
private data class StremioSubtitleSurrogate(
    val id: kotlinx.serialization.json.JsonElement? = null,
    val lang: String? = null,
    val label: String? = null,
    val url: String? = null,
    val name: String? = null,
    val title: String? = null,
)

@Serializable
private data class StremioCatalogExtraSurrogate(
    val name: String = "",
    @SerialName("isRequired") val isRequired: Boolean = false,
    val options: List<String> = emptyList(),
    @SerialName("optionsLimit") val optionsLimit: Int? = null,
) {
    fun toCatalogExtra(): StremioCatalogExtra = StremioCatalogExtra(
        name = name,
        isRequired = isRequired,
        options = options,
        optionsLimit = optionsLimit,
    )
}

fun StremioStream.qualityScore(): Double {
    val haystack = listOfNotNull(
        name,
        title,
        description,
        behaviorHints?.filename,
    ).joinToString(" ").lowercase()

    val base = qualityPatterns.firstOrNull { (needle, _) -> haystack.contains(needle) }?.second ?: 0.50
    val hdrBoost = if (haystack.contains("hdr") || haystack.contains("dolby vision") || haystack.contains("dv")) 0.04 else 0.0
    val remuxBoost = if (haystack.contains("remux") || haystack.contains("bluray")) 0.04 else 0.0
    val webBoost = if (haystack.contains("web-dl") || haystack.contains("webrip")) 0.02 else 0.0
    val notWebReadyPenalty = if (behaviorHints?.notWebReady == true) 0.08 else 0.0

    return (base + hdrBoost + remuxBoost + webBoost - notWebReadyPenalty).coerceIn(0.0, 1.0)
}


